plugins/serialization/src/main/java/org/qbicc/plugin/serialization/InstanceInterceptor.java

```diff
@@ -0,0 +1,49 @@
+package org.qbicc.plugin.serialization;
+
+import static org.qbicc.graph.atomic.AccessModes.SingleUnshared;
+
+import org.qbicc.context.CompilationContext;
+import org.qbicc.interpreter.VmObject;
+import org.qbicc.type.definition.LoadedTypeDefinition;
+import org.qbicc.type.descriptor.ClassTypeDescriptor;
+
+/**
+ * This class enables VmObject instances to be modified
+ * immediately before they are serialized.
+ */
+class InstanceInterceptor {
+    private final CompilationContext ctxt;
+
+    InstanceInterceptor(CompilationContext ctxt) {
+        this.ctxt = ctxt;
+    }
+
+     VmObject process(VmObject original) {
+        ClassTypeDescriptor otd = original.getVmClass().getTypeDefinition().getDescriptor();
+        if (otd.packageAndClassNameEquals("java/io", "FileDescriptor")) {
+            return interceptFileDescriptor(original);
+        }
+
+        return original;
+    }
+
+
+    private VmObject interceptFileDescriptor(VmObject original) {
+        LoadedTypeDefinition fdType = original.getVmClass().getTypeDefinition();
+
+        // Do not change already closed descriptors and the special descriptors stdin, stdout, stderr.
+        if (original.getMemory().load32(original.indexOf(fdType.findField("fd")), SingleUnshared) < 3) {
+            return original;
+        }
```

Comment:
I don't think we can or should preserve the illusion that FDs 0..2 are preserved between build time and run time.  We will definitely need to re-open them at run time, probably using a run time initializer.  So I think it would be best to just close them all.

Reply:
I can try it, but I suspect that will actually break `System.out.println` until we get the runtime initialization working.

Reply:
Yeah I think that would be expected. We'd need a run time aspect for `FileDescriptor` and then a resolution for #950 which includes a run time init sequence that resets `System.in/out/err` (which should be reset to `null` by a run time aspect initializer for those fields, and which in turn should prevent those fields' values from being serialized in the first place).